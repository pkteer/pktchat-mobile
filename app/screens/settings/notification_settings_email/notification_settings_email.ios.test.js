// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {shallow} from 'enzyme';
import React from 'react';

import Preferences from '@mm-redux/constants/preferences';
import SectionItem from '@screens/settings/section_item';

import NotificationSettingsEmailIos from './notification_settings_email.ios.js';

jest.mock('@utils/theme', () => {
    const original = jest.requireActual('../../../utils/theme');
    return {
        ...original,
        changeOpacity: jest.fn(),
    };
});

describe('NotificationSettingsEmailIos', () => {
    const baseProps = {
        currentUser: {id: 'current_user_id'},
        notifyProps: {
            email: 'true',
            email_threads: 'all',
        },
        emailInterval: '30',
        enableEmailBatching: false,
        actions: {
            updateMe: jest.fn(),
            savePreferences: jest.fn(),
        },
        sendEmailNotifications: true,
        theme: Preferences.THEMES.denim,
        componentId: 'component-id',
        isCollapsedThreadsEnabled: false,
    };

    test('should match snapshot, renderEmailSection', () => {
        const wrapper = shallow(
            <NotificationSettingsEmailIos {...baseProps}/>,
        );

        expect(wrapper.instance().renderEmailSection()).toMatchSnapshot();
    });

    test('should match snapshot, renderEmailThreadsSection', () => {
        const style = {
            separator: {},
        };

        const wrapper = shallow(
            <NotificationSettingsEmailIos {...{...baseProps, isCollapsedThreadsEnabled: true}}/>,
        );

        expect(wrapper.instance().renderEmailThreadsSection(style)).toMatchSnapshot();
    });

    test('should save preference on back button only if email interval has changed', () => {
        const wrapper = shallow(
            <NotificationSettingsEmailIos {...baseProps}/>,
        );

        const instance = wrapper.instance();

        // should not save preference if email interval has not changed.
        instance.componentDidDisappear();
        expect(baseProps.actions.updateMe).toHaveBeenCalledTimes(0);
        expect(baseProps.actions.savePreferences).toHaveBeenCalledTimes(0);

        // should save preference if email interval has changed.
        wrapper.setState({newInterval: '0'});
        instance.componentDidDisappear();
        expect(baseProps.actions.updateMe).toHaveBeenCalledTimes(1);
        expect(baseProps.actions.savePreferences).toHaveBeenCalledTimes(1);
    });

    test('should call actions.updateMe and actions.savePreferences on saveEmailNotifyProps', () => {
        const savePreferences = jest.fn();
        const updateMe = jest.fn();
        const props = {...baseProps, actions: {savePreferences, updateMe}};
        const wrapper = shallow(
            <NotificationSettingsEmailIos {...props}/>,
        );

        wrapper.setState({email: 'true', newInterval: 30});
        wrapper.instance().saveEmailNotifyProps();

        expect(updateMe).toHaveBeenCalledTimes(1);
        expect(updateMe.mock.calls[0][0].notify_props.email).toBe('true');

        expect(savePreferences).toHaveBeenCalledTimes(1);
        expect(savePreferences).toBeCalledWith('current_user_id', [{category: 'notifications', name: 'email_interval', user_id: 'current_user_id', value: 30}]);
    });

    test('should match state on setEmailInterval', () => {
        const wrapper = shallow(
            <NotificationSettingsEmailIos {...baseProps}/>,
        );

        wrapper.setState({interval: '0'});
        wrapper.instance().setEmailInterval('30');
        expect(wrapper.state({interval: '30'}));

        wrapper.instance().setEmailInterval('0');
        expect(wrapper.state({interval: '0'}));

        wrapper.instance().setEmailInterval('3600');
        expect(wrapper.state({interval: '3600'}));
    });

    test('should match state on action of SectionItem', () => {
        const wrapper = shallow(
            <NotificationSettingsEmailIos
                {...baseProps}
                sendEmailNotifications={false}
                enableEmailBatching={false}
            />,
        );

        expect(wrapper.find(SectionItem).exists()).toBe(false);

        wrapper.setProps({sendEmailNotifications: true});
        expect(wrapper.find(SectionItem).exists()).toBe(true);
        expect(wrapper.find(SectionItem).length).toBe(2);

        wrapper.setProps({enableEmailBatching: true});
        expect(wrapper.find(SectionItem).exists()).toBe(true);
        expect(wrapper.find(SectionItem).length).toBe(4);

        wrapper.setState({email: 'false', interval: '0'});

        wrapper.find(SectionItem).first().prop('action')('30');
        expect(wrapper.state({email: 'true', interval: '30'}));

        wrapper.find(SectionItem).first().prop('action')('0');
        expect(wrapper.state({email: 'true', interval: '0'}));

        wrapper.find(SectionItem).last().prop('action')('3600');
        expect(wrapper.state({email: 'true', interval: '3600'}));
    });

    test('should call props.actions.savePreferences on saveUserNotifyProps', () => {
        const props = {...baseProps, actions: {savePreferences: jest.fn(), updateMe: jest.fn()}};
        const wrapper = shallow(
            <NotificationSettingsEmailIos {...props}/>,
        );

        wrapper.setState({email: 'true', newInterval: '3600'});
        wrapper.instance().saveEmailNotifyProps();
        expect(props.actions.savePreferences).toHaveBeenCalledTimes(1);
        expect(props.actions.savePreferences).toBeCalledWith(
            'current_user_id',
            [{category: 'notifications', name: 'email_interval', user_id: 'current_user_id', value: '3600'}],
        );
    });
});
