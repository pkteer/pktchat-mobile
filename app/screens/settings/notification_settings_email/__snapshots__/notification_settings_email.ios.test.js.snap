// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationSettingsEmailIos should match snapshot, renderEmailSection 1`] = `
<section
  disableFooter={false}
  footerDefaultMessage="Email notifications are sent for mentions and direct messages when you are offline or away for more than 5 minutes."
  footerId="user.settings.notifications.emailInfo"
  headerDefaultMessage="SEND EMAIL NOTIFICATIONS"
  headerId="mobile.notification_settings.email.send"
  theme={
    Object {
      "awayIndicator": "#ffbc1f",
      "buttonBg": "#1c58d9",
      "buttonColor": "#ffffff",
      "centerChannelBg": "#ffffff",
      "centerChannelColor": "#3f4350",
      "codeTheme": "github",
      "dndIndicator": "#d24b4e",
      "errorTextColor": "#d24b4e",
      "linkColor": "#386fe5",
      "mentionBg": "#ffffff",
      "mentionColor": "#1e325c",
      "mentionHighlightBg": "#ffd470",
      "mentionHighlightLink": "#1b1d22",
      "newMessageSeparator": "#cc8f00",
      "onlineIndicator": "#3db887",
      "sidebarBg": "#1e325c",
      "sidebarHeaderBg": "#192a4d",
      "sidebarHeaderTextColor": "#ffffff",
      "sidebarTeamBarBg": "#14213e",
      "sidebarText": "#ffffff",
      "sidebarTextActiveBorder": "#5d89ea",
      "sidebarTextActiveColor": "#ffffff",
      "sidebarTextHoverBg": "#28427b",
      "sidebarUnreadText": "#ffffff",
      "type": "Denim",
    }
  }
>
  <View>
    <sectionItem
      action={[Function]}
      actionType="select"
      actionValue="30"
      label={
        <InjectIntl(FormattedText)
          defaultMessage="Immediately"
          id="user.settings.notifications.email.immediately"
        />
      }
      selected={true}
      testID="notification_settings_email.immediately.action"
      theme={
        Object {
          "awayIndicator": "#ffbc1f",
          "buttonBg": "#1c58d9",
          "buttonColor": "#ffffff",
          "centerChannelBg": "#ffffff",
          "centerChannelColor": "#3f4350",
          "codeTheme": "github",
          "dndIndicator": "#d24b4e",
          "errorTextColor": "#d24b4e",
          "linkColor": "#386fe5",
          "mentionBg": "#ffffff",
          "mentionColor": "#1e325c",
          "mentionHighlightBg": "#ffd470",
          "mentionHighlightLink": "#1b1d22",
          "newMessageSeparator": "#cc8f00",
          "onlineIndicator": "#3db887",
          "sidebarBg": "#1e325c",
          "sidebarHeaderBg": "#192a4d",
          "sidebarHeaderTextColor": "#ffffff",
          "sidebarTeamBarBg": "#14213e",
          "sidebarText": "#ffffff",
          "sidebarTextActiveBorder": "#5d89ea",
          "sidebarTextActiveColor": "#ffffff",
          "sidebarTextHoverBg": "#28427b",
          "sidebarUnreadText": "#ffffff",
          "type": "Denim",
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": undefined,
          "flex": 1,
          "height": 1,
          "marginLeft": 15,
        }
      }
    />
    <sectionItem
      action={[Function]}
      actionType="select"
      actionValue="0"
      label={
        <InjectIntl(FormattedText)
          defaultMessage="Never"
          id="user.settings.notifications.email.never"
        />
      }
      selected={false}
      testID="notification_settings_email.never.action"
      theme={
        Object {
          "awayIndicator": "#ffbc1f",
          "buttonBg": "#1c58d9",
          "buttonColor": "#ffffff",
          "centerChannelBg": "#ffffff",
          "centerChannelColor": "#3f4350",
          "codeTheme": "github",
          "dndIndicator": "#d24b4e",
          "errorTextColor": "#d24b4e",
          "linkColor": "#386fe5",
          "mentionBg": "#ffffff",
          "mentionColor": "#1e325c",
          "mentionHighlightBg": "#ffd470",
          "mentionHighlightLink": "#1b1d22",
          "newMessageSeparator": "#cc8f00",
          "onlineIndicator": "#3db887",
          "sidebarBg": "#1e325c",
          "sidebarHeaderBg": "#192a4d",
          "sidebarHeaderTextColor": "#ffffff",
          "sidebarTeamBarBg": "#14213e",
          "sidebarText": "#ffffff",
          "sidebarTextActiveBorder": "#5d89ea",
          "sidebarTextActiveColor": "#ffffff",
          "sidebarTextHoverBg": "#28427b",
          "sidebarUnreadText": "#ffffff",
          "type": "Denim",
        }
      }
    />
  </View>
</section>
`;

exports[`NotificationSettingsEmailIos should match snapshot, renderEmailThreadsSection 1`] = `
<section
  footerDefaultMessage="When enabled, any reply to a thread you're following will send an email notification."
  footerId="user.settings.notifications.email_threads.info"
  headerDefaultMessage="THREAD REPLY NOTIFICATIONS"
  headerId="user.settings.notifications.email_threads.title"
  theme={
    Object {
      "awayIndicator": "#ffbc1f",
      "buttonBg": "#1c58d9",
      "buttonColor": "#ffffff",
      "centerChannelBg": "#ffffff",
      "centerChannelColor": "#3f4350",
      "codeTheme": "github",
      "dndIndicator": "#d24b4e",
      "errorTextColor": "#d24b4e",
      "linkColor": "#386fe5",
      "mentionBg": "#ffffff",
      "mentionColor": "#1e325c",
      "mentionHighlightBg": "#ffd470",
      "mentionHighlightLink": "#1b1d22",
      "newMessageSeparator": "#cc8f00",
      "onlineIndicator": "#3db887",
      "sidebarBg": "#1e325c",
      "sidebarHeaderBg": "#192a4d",
      "sidebarHeaderTextColor": "#ffffff",
      "sidebarTeamBarBg": "#14213e",
      "sidebarText": "#ffffff",
      "sidebarTextActiveBorder": "#5d89ea",
      "sidebarTextActiveColor": "#ffffff",
      "sidebarTextHoverBg": "#28427b",
      "sidebarUnreadText": "#ffffff",
      "type": "Denim",
    }
  }
>
  <sectionItem
    action={[Function]}
    actionType="toggle"
    description={<View />}
    label={
      <InjectIntl(FormattedText)
        defaultMessage="Notify me about all replies to threads I'm following"
        id="user.settings.notifications.email_threads.description"
      />
    }
    selected={true}
    theme={
      Object {
        "awayIndicator": "#ffbc1f",
        "buttonBg": "#1c58d9",
        "buttonColor": "#ffffff",
        "centerChannelBg": "#ffffff",
        "centerChannelColor": "#3f4350",
        "codeTheme": "github",
        "dndIndicator": "#d24b4e",
        "errorTextColor": "#d24b4e",
        "linkColor": "#386fe5",
        "mentionBg": "#ffffff",
        "mentionColor": "#1e325c",
        "mentionHighlightBg": "#ffd470",
        "mentionHighlightLink": "#1b1d22",
        "newMessageSeparator": "#cc8f00",
        "onlineIndicator": "#3db887",
        "sidebarBg": "#1e325c",
        "sidebarHeaderBg": "#192a4d",
        "sidebarHeaderTextColor": "#ffffff",
        "sidebarTeamBarBg": "#14213e",
        "sidebarText": "#ffffff",
        "sidebarTextActiveBorder": "#5d89ea",
        "sidebarTextActiveColor": "#ffffff",
        "sidebarTextHoverBg": "#28427b",
        "sidebarUnreadText": "#ffffff",
        "type": "Denim",
      }
    }
  />
  <View
    style={Object {}}
  />
</section>
`;
